File: ngeoltx/src/Ngeoltx.Engine/Validator.cs
using System.Text.RegularExpressions;

namespace Ngeoltx.Engine;

public sealed record Diagnostic(string Severity, int? Row, string Message, string Detail = "")
{
    public const string Error = "error";
    public const string Warning = "warning";

    public bool IsError => Severity == Error;

    public override string ToString() =>
        (Row is null ? "program" : "row " + Row) + ": " + Message;
}

public sealed class ValidationReport : List<Diagnostic>
{
    public IReadOnlyList<Diagnostic> Errors => this.Where(d => d.IsError).ToList();
    public IReadOnlyList<Diagnostic> Warnings =>
        this.Where(d => d.Severity == Diagnostic.Warning).ToList();

    public bool Ok => Errors.Count == 0;

    public string Summary()
    {
        if (Ok && Warnings.Count == 0) return "Program valid.";
        var bits = new List<string>();
        if (Errors.Count > 0) bits.Add(Errors.Count + " error(s)");
        if (Warnings.Count > 0) bits.Add(Warnings.Count + " warning(s)");
        return string.Join(", ", bits);
    }
}

/// <summary>
/// Static checks run before a program is allowed to execute.
/// </summary>
/// <remarks>
/// The single highest-value addition over v1. There, a mistyped verb became an
/// AttributeError raised from getattr at row 200, with the supply already at
/// 13.5 V and relays closed. Everything checkable here is checked while the
/// fixture is still cold, and the UI refuses to arm Run until it is clean.
/// </remarks>
public static class Validator
{
    /// <summary>Verbs whose column-2 argument names a label.</summary>
    private static readonly HashSet<string> JumpVerbs = new()
    {
        "J", "JC", "JCM", "JNC", "JE", "JNE", "JET", "JNET", "JG", "JL", "JGE", "JLE",
    };

    private static readonly Regex ReferencePattern = new(@"^\*(.+)$", RegexOptions.Compiled);
    private static readonly Regex CoordPattern =
        new(@"^\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*$", RegexOptions.Compiled);

    public static ValidationReport Validate(TestProgram program, VerbRegistry? registry = null)
    {
        registry ??= VerbRegistry.Default;
        var report = new ValidationReport();

        void Error(int? row, string message, string detail = "") =>
            report.Add(new Diagnostic(Diagnostic.Error, row, message, detail));
        void Warn(int? row, string message, string detail = "") =>
            report.Add(new Diagnostic(Diagnostic.Warning, row, message, detail));

        // -- structure ---------------------------------------------------
        if (program.SectionOrNull("Exec") is null)
            Error(null, "program has no <Exec> section -- nothing would run");
        if (program.Modules.Count == 0)
            Warn(null, "program declares no <Modules>; only built-in verbs will resolve");
        if (!program.HasTeardown)
            Warn(null, "program has no <Teardown> section",
                 "Without one, an aborted run leaves the fixture in whatever state it "
                 + "reached. Add <Teardown> to power down supplies and open relays.");

        // -- modules -----------------------------------------------------
        var knownModules = registry.Modules().ToHashSet();
        foreach (var (alias, target) in program.Modules)
            if (!knownModules.Contains(registry.ResolveModule(target)))
                Error(null,
                    "module alias '" + alias + "' -> '" + target
                    + "' is not a registered driver",
                    "Registered drivers: " + string.Join(", ", knownModules.OrderBy(m => m)));

        var aliveSize = DeclaredAliveSize(program);
        var dims = DeclaredDims(program);

        // -- rows --------------------------------------------------------
        foreach (var row in program.Rows)
        {
            if (row.IsBlank || row.IsMarker || row.Module == "-") continue;
            if (program.InDeclarativeSection(row.Index)) continue;

            if (row.Module.Length == 0)
            {
                // A warning, not an error: the row is skipped and the program
                // still loads. v1 raised KeyError on globals[""] and diverted to
                // the row's handler, so the step never ran there either --
                // blocking the load would be stricter than what this replaces.
                if (row.Verb.Length > 0)
                    Warn(row.Index,
                         "'" + row.Verb + "' has no module in column 0 -- row will be skipped",
                         "Fill in column 0 to make this step run. In v1 it raised KeyError "
                         + "and diverted to the row's exception handler, so it never "
                         + "executed there either.");
                continue;
            }

            var module = program.Modules.TryGetValue(row.Module, out var mapped)
                ? mapped : row.Module;
            var spec = registry.Lookup(module, row.Verb);
            if (spec is null)
            {
                if (!knownModules.Contains(registry.ResolveModule(module)))
                    Error(row.Index, "unknown module '" + row.Module + "'");
                else
                {
                    var near = VerbRegistry.Suggest(row.Verb, registry.VerbsFor(module));
                    Error(row.Index, "unknown verb " + row.Module + "." + row.Verb
                        + (near is null ? "" : " -- did you mean " + near + "?"));
                }
                continue;
            }

            foreach (var param in spec.Params.Where(p => p.Required))
                if (!row.Has(param.Index))
                    Error(row.Index,
                        row.Module + "." + row.Verb + ": missing required argument '"
                        + param.Name + "' (column " + param.Index + ")", param.Doc);

            // Arguments beyond what the verb declares are nearly always a
            // copy-paste slip from a neighbouring row.
            if (spec.Params.Count > 0)
                for (var i = Row.ArgStart; i <= Row.ArgEnd; i++)
                    if (row.Has(i) && spec.ParamAt(i) is null)
                        Warn(row.Index,
                            row.Module + "." + row.Verb + ": unexpected argument in column "
                            + i + " ('" + row.Raw(i) + "') -- this verb ignores it");

            if (JumpVerbs.Contains(row.Verb) && row.Has(2))
            {
                var target = row.Raw(2);
                if (!target.StartsWith("*") && !program.Labels.ContainsKey(target))
                {
                    var near = VerbRegistry.Suggest(target, program.Labels.Keys.ToList());
                    Error(row.Index, "jump to undefined label '" + target + "'"
                        + (near is null ? "" : " -- did you mean " + near + "?"));
                }
            }

            var route = row.Route;
            if (route is not null && !program.Labels.ContainsKey(route) && route != "STD_EX")
                Warn(row.Index,
                     "exception label '" + route + "' has no matching LABEL row",
                     "On failure the run will abort instead of recovering.");

            if (aliveSize is not null)
                foreach (var token in row.Alive.Split(',').Select(t => t.Trim()))
                {
                    if (token.Length == 0 || token == "-") continue;
                    if (!int.TryParse(token, out var index))
                        Error(row.Index, "bad alive mask entry '" + token + "'");
                    else if (index >= aliveSize)
                        Error(row.Index, "alive mask references UUT " + token
                            + " but initAlive declares " + aliveSize);
                }

            for (var i = Row.ArgStart; i <= Row.ArgEnd; i++)
            {
                var raw = row.Raw(i);
                var match = ReferencePattern.Match(raw);
                if (!match.Success) continue;

                var name = match.Groups[1].Value;
                var reference = program.Vars.TryGetValue(name, out var coord) ? coord : name;
                var parsed = CoordPattern.Match(reference);
                if (!parsed.Success)
                {
                    var near = VerbRegistry.Suggest(name, program.Vars.Keys.ToList());
                    Error(row.Index, "column " + i + ": '" + raw
                        + "' is neither an l,c,p coordinate nor a declared variable"
                        + (near is null ? "" : " -- did you mean *" + near + "?"));
                    continue;
                }
                if (dims is null) continue;
                var (l, c, p) = (int.Parse(parsed.Groups[1].Value),
                                 int.Parse(parsed.Groups[2].Value),
                                 int.Parse(parsed.Groups[3].Value));
                if (l >= dims.Value.Lines || c >= dims.Value.Cols || p >= dims.Value.Pages)
                    Error(row.Index, "column " + i + ": " + raw
                        + " is outside the data store declared by INITDATA ("
                        + dims.Value.Lines + "x" + dims.Value.Cols + "x"
                        + dims.Value.Pages + ")");
            }
        }

        CheckParallelBlocks(program, report);
        return report;
    }

    /// <summary>Every parallel block must be closed, and nesting is unsupported.</summary>
    private static void CheckParallelBlocks(TestProgram program, ValidationReport report)
    {
        int? openAt = null;
        foreach (var row in program.Rows)
        {
            var token = row.Module.Trim();
            if (token == Sequencer.ParallelOpen)
            {
                if (openAt is not null)
                    report.Add(new Diagnostic(Diagnostic.Error, row.Index,
                        "<ParallelTask> nested inside another parallel block",
                        "Outer block opened at row " + openAt + "."));
                openAt = row.Index;
            }
            else if (token == Sequencer.ParallelClose)
            {
                if (openAt is null)
                    report.Add(new Diagnostic(Diagnostic.Error, row.Index,
                        "<ParallelTask/> closes a block that was never opened"));
                openAt = null;
            }
        }
        if (openAt is not null)
            report.Add(new Diagnostic(Diagnostic.Error, openAt,
                "<ParallelTask> is never closed",
                "v1 would run off the end of the table here."));
    }

    public static int? DeclaredAliveSize(TestProgram program)
    {
        foreach (var row in program.Rows)
            if (row.Verb == "initAlive" && row.Has(2)
                && int.TryParse(row.Raw(2), out var count))
                return count;
        return null;
    }

    public static (int Lines, int Cols, int Pages)? DeclaredDims(TestProgram program)
    {
        foreach (var row in program.Rows)
        {
            if (row.Verb != "INITDATA") continue;
            if (int.TryParse(row.Raw(2), out var l) && int.TryParse(row.Raw(3), out var c)
                && int.TryParse(row.Raw(4), out var p))
                return (l, c, p);
        }
        return null;
    }
}

File: ngwart/ngwart/engine/validator.py
"""Static checks run before a program is allowed to execute.

This is the single highest-value addition in v2. In v1 a mistyped verb became

    AttributeError: module 'WinSerialManager' has no attribute 'EXCHANGBYTES_LVS'

raised from ``getattr`` at ngine.py:285 -- discovered at row 200, with the PSU
already at 13.5 V and relays closed. Everything checkable here is checked while
the fixture is still cold, and the UI refuses to arm Play until it is clean.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from .program import ARG_END, ARG_START, Program
from .registry import REGISTRY, Registry

#: Verbs whose column-2 argument names a label.
_JUMP_VERBS = {
    "J", "JC", "JCM", "JNC", "JE", "JNE", "JET", "JNET", "JG", "JL", "JGE", "JLE",
}

_REF_RE = re.compile(r"^\*(.+)$")
_COORD_RE = re.compile(r"^\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*$")

ERROR = "error"
WARNING = "warning"


@dataclass(frozen=True)
class Diagnostic:
    severity: str
    row: int | None
    message: str
    detail: str = ""

    @property
    def is_error(self) -> bool:
        return self.severity == ERROR

    def __str__(self) -> str:
        where = f"row {self.row}" if self.row is not None else "program"
        return f"{self.severity.upper():7} {where}: {self.message}"


class Report(list):
    """A list of Diagnostics that knows whether it blocks execution."""

    @property
    def errors(self) -> list[Diagnostic]:
        return [d for d in self if d.is_error]

    @property
    def warnings(self) -> list[Diagnostic]:
        return [d for d in self if d.severity == WARNING]

    @property
    def ok(self) -> bool:
        return not self.errors

    def summary(self) -> str:
        if self.ok and not self.warnings:
            return "Program valid."
        bits = []
        if self.errors:
            bits.append(f"{len(self.errors)} error(s)")
        if self.warnings:
            bits.append(f"{len(self.warnings)} warning(s)")
        return ", ".join(bits)


def validate(program: Program, registry: Registry | None = None) -> Report:
    reg = registry or REGISTRY
    report = Report()

    def err(row, msg, detail=""):
        report.append(Diagnostic(ERROR, row, msg, detail))

    def warn(row, msg, detail=""):
        report.append(Diagnostic(WARNING, row, msg, detail))

    # -- structure --------------------------------------------------------
    if program.section("Exec") is None:
        err(None, "program has no <Exec> section -- nothing would run")

    # -- <values> ---------------------------------------------------------
    # An error, not a warning: a variable whose value never arrived surfaces
    # much later as "bad data reference ''" or as a test that silently measured
    # the wrong place. The whole point of reading the file at load time is to
    # find this while the fixture is still cold.
    for problem in getattr(program, "value_problems", []):
        err(None, problem,
            "Variables take their value from the <values> file named in "
            "<Vars>. Re-run `ngwart teach` to regenerate it, or fix the key.")
    unclaimed = set(program.values_loaded) - set(program.var_sources.values())
    if unclaimed:
        listed = ", ".join(sorted(unclaimed)[:6])
        warn(None,
             f"{len(unclaimed)} key(s) in the values file match no variable: "
             f"{listed}",
             "Harmless, but usually a renamed variable that left its value "
             "behind.")
    if not program.modules:
        warn(None, "program declares no <Modules>; only built-in verbs will resolve")
    if not program.has_teardown:
        warn(None, "program has no <Teardown> section",
             "Without one, an aborted run leaves the fixture in whatever state it "
             "reached. Add <Teardown> to power down supplies and open relays.")

    # -- modules ----------------------------------------------------------
    known_modules = set(reg.modules())
    for alias, target in program.modules.items():
        if reg.resolve_module(target) not in known_modules:
            err(None, f"module alias '{alias}' -> '{target}' is not a registered driver",
                f"Registered drivers: {', '.join(sorted(known_modules))}")

    alive_size = _declared_alive_size(program)
    dims = _declared_dims(program)

    # -- rows -------------------------------------------------------------
    for row in program.rows:
        if row.is_blank or row.is_marker or row.module == "-":
            continue
        # <Modules> and <Vars> hold declarations, not instructions.
        if program.in_declarative_section(row.index):
            continue
        if not row.module:
            if row.verb:
                # A warning, not an error: the row is skipped and the program
                # still loads. v1 raised KeyError on globals()[""] and diverted
                # to the row's handler, so the step never ran there either --
                # blocking the load would be stricter than the behaviour this
                # replaces. The warning stays so the skip is never silent.
                warn(row.index,
                     f"'{row.verb}' has no module in column 0 -- row will be skipped",
                     "Fill in column 0 to make this step run. In v1 it raised "
                     "KeyError and diverted to the row's exception handler, so "
                     "it never executed there either.")
            continue

        module = program.modules.get(row.module, row.module)
        spec = reg.lookup(module, row.verb)
        if spec is None:
            if reg.resolve_module(module) not in known_modules:
                err(row.index, f"unknown module '{row.module}'")
            else:
                near = _suggest(row.verb, reg.verbs_for(module))
                hint = f" -- did you mean {near}?" if near else ""
                err(row.index, f"unknown verb {row.module}.{row.verb}{hint}")
            continue

        # arity
        for param in spec.params:
            if param.required and not row.has(param.index):
                err(row.index,
                    f"{row.module}.{row.verb}: missing required argument "
                    f"'{param.name}' (column {param.index})",
                    param.doc)

        # arguments beyond what the verb declares are almost always a
        # copy-paste slip from a neighbouring row
        if spec.params:
            for i in range(ARG_START, ARG_END + 1):
                if row.has(i) and spec.param_at(i) is None:
                    warn(row.index,
                         f"{row.module}.{row.verb}: unexpected argument in column {i} "
                         f"('{row.raw(i)}') -- this verb ignores it")

        # jump targets
        if row.verb in _JUMP_VERBS and row.has(2):
            target = row.raw(2)
            if not target.startswith("*") and target not in program.labels:
                near = _suggest(target, list(program.labels))
                hint = f" -- did you mean {near}?" if near else ""
                err(row.index, f"jump to undefined label '{target}'{hint}")

        # exception routing target
        route = row.route
        if route and route not in program.labels and route != "STD_EX":
            warn(row.index,
                 f"exception label '{route}' has no matching LABEL row",
                 "On failure the run will abort instead of recovering.")

        # alive mask
        if alive_size is not None:
            for token in [t.strip() for t in row.alive.split(",")]:
                if token in ("", "-"):
                    continue
                if not token.isdigit():
                    err(row.index, f"bad alive mask entry '{token}'")
                elif int(token) >= alive_size:
                    err(row.index,
                        f"alive mask references UUT {token} but initAlive "
                        f"declares {alive_size}")

        # data references
        for i in range(ARG_START, ARG_END + 1):
            raw = row.raw(i)
            m = _REF_RE.match(raw)
            if not m:
                continue
            ref = program.vars.get(m.group(1), m.group(1))
            coord = _COORD_RE.match(ref)
            if not coord:
                near = _suggest(m.group(1), list(program.vars))
                hint = f" -- did you mean *{near}?" if near else ""
                err(row.index,
                    f"column {i}: '{raw}' is neither an l,c,p coordinate nor a "
                    f"declared variable{hint}")
                continue
            if dims:
                l, c, p = (int(coord.group(1)), int(coord.group(2)), int(coord.group(3)))
                if l >= dims[0] or c >= dims[1] or p >= dims[2]:
                    err(row.index,
                        f"column {i}: {raw} is outside the data store declared by "
                        f"INITDATA ({dims[0]}x{dims[1]}x{dims[2]})")

    _check_parallel_blocks(program, report)
    return report


def _check_parallel_blocks(program: Program, report: Report) -> None:
    """Every <ParallelTask> must be closed, and nesting is not supported."""
    open_at: int | None = None
    for row in program.rows:
        tok = row.module.strip()
        if tok == "<ParallelTask>":
            if open_at is not None:
                report.append(Diagnostic(
                    ERROR, row.index,
                    "<ParallelTask> nested inside another parallel block",
                    f"Outer block opened at row {open_at}."))
            open_at = row.index
        elif tok == "<ParallelTask/>":
            if open_at is None:
                report.append(Diagnostic(
                    ERROR, row.index, "<ParallelTask/> closes a block that was never opened"))
            open_at = None
    if open_at is not None:
        report.append(Diagnostic(
            ERROR, open_at,
            "<ParallelTask> is never closed",
            "v1 would run off the end of the table here."))


def _declared_alive_size(program: Program) -> int | None:
    for row in program.rows:
        if row.verb == "initAlive" and row.has(2) and row.raw(2).isdigit():
            return int(row.raw(2))
    return None


def _declared_dims(program: Program) -> tuple[int, int, int] | None:
    for row in program.rows:
        if row.verb == "INITDATA":
            vals = [row.raw(i) for i in (2, 3, 4)]
            if all(v.isdigit() for v in vals):
                return tuple(int(v) for v in vals)  # type: ignore[return-value]
    return None


def _suggest(word: str, candidates: list[str]) -> str | None:
    import difflib

    hits = difflib.get_close_matches(word, [c for c in candidates if c], n=1, cutoff=0.7)
    return hits[0] if hits else None
